packages/node/src/policy-gen/load-for-policy.js

```diff
@@ -0,0 +1,463 @@
+  modulesToInspect,
+  globalsForPackage,
+  builtinsForPackage,
+  violationsForPackage,
+  pendingInspections
+) => {
+  let messageCount = 0
+
+  /**
+   * @param {Omit<
+   *   LocalModuleSource,
+   *   'location' | 'sourceLocation' | 'sourceDirname' | 'parser' | 'record'
+   * > & { language: Language; location: FileUrlString }} moduleSource
+   * @param {CanonicalName} canonicalName
+   * @returns {void}
+   */
+  const inspectModuleSource = (moduleSource, canonicalName) => {
+    const { bytes: source, location: id, language } = moduleSource
+    const type = MessageTypes.Inspect
+    /** @type {SourceType | undefined} */
+    let sourceType
+    switch (language) {
+      case LANGUAGE_MJS:
+        sourceType = SOURCE_TYPE_MODULE
+        break
+      case LANGUAGE_CJS:
+        sourceType = SOURCE_TYPE_SCRIPT
+        break
+      default: {
+        log.debug(
+          `Unsupported language "${language}" for module ${id}; skipping`
+        )
+        return
+      }
+    }
+
+    /** @type {InspectMessage} */
+    const message = {
+      source,
+      id,
+      sourceType,
+      type,
+    }
+
+    log.debug(`Inspecting module: ${id}`)
+
+    const inspectionPromise = workerPool
+      .sendTask(message, MessageTypes.Policies)
+      .then((message) => {
+        inspectedModules.add(id)
+        if (process.stderr.isTTY) {
+          messageCount++
+          const trianglePos = ((messageCount - 1) % 3) + 1
+          const prefix = '   '.split('')
+
+          // Style the triangle based on position
+          let styledTriangle
+          if (trianglePos === 1) {
+            styledTriangle = chalk.dim('▶')
+          } else if (trianglePos === 2) {
+            styledTriangle = chalk.white('▶')
+          } else {
+            styledTriangle = chalk.whiteBright('▶')
+          }
+
+          prefix[trianglePos - 1] = styledTriangle
+          const prefixStr = prefix.join('')
+          process.stderr.write(
+            `\r        ${chalk.dim('›')} ${prefixStr} ${chalk.white('Inspecting modules: ')}${chalk.whiteBright(inspectedModules.size)}${chalk.dim('/')}${chalk.white(modulesToInspect.size)}`
+          )
+        }
+        if (message.type === MessageTypes.Policies) {
+          const { globalPolicy, builtinPolicy, violations } = message
+          if (globalPolicy) {
+            const currentGlobalPolicy = globalsForPackage.get(canonicalName)
+            globalsForPackage.set(canonicalName, {
+              ...currentGlobalPolicy,
+              ...globalPolicy,
+            })
+          }
+          if (builtinPolicy) {
+            const currentBuiltinPolicy = builtinsForPackage.get(canonicalName)
+            builtinsForPackage.set(canonicalName, {
+              ...currentBuiltinPolicy,
+              ...builtinPolicy,
+            })
+          }
+          if (violations) {
+            const currentViolations = violationsForPackage.get(canonicalName)
+            if (currentViolations) {
```

Comment:
(architecture) this function mixes come very distinct responsibilities. It understands messages from workers, displaying progress but also structure of policy and types of violations. 
If we add a new violation, I don't want to go and update this function. I'd rather generalize the aggregation and/or mke it happen elsewhere.

I'd separate out all of the worker stuff from logic and have inspector-inner and inspector-outer code that understands what it's handling, with the worker stuff inbetween. That separation also means that the part that runs inside the worker is the scope that belongs in core.

Does it all have to happen in this PR? probably not
But we at the least need a well written issue for that.

Reply:
By "elsewhere" you mean in the parent process somewhere?  Or "elsewhere" in some separate function?  I agree it could be split up.